1)	Peguem uma empresa que respondeu menos e façam o seguinte:
-	Filtrar as reclamações que a empresa não respondeu
-	A IA irá traduzir as reclamações para Português
-	A IA vai fazer um resumo da reclamação
-	A IA vai detectar o nível de risco do cliente (BAIXO, MEDIO, ALTO)
Risco Alto: Palavras que ameaça diretamente a empresa com processo, advogado, linguagem agressiva ou tóxica. Ameaça deixar de utilizar os produtos da empresa.
Risco Médio: Cliente não ameaça diretamente a empresa mas utiliza algumas palavras agressivas demonstrando sinal de estresse mas não dá indícios de deixar de consumir algum produto da empresa.
Risco Baixo: Cliente faz a reclamação sem ameaças a empresa e demonstra insatisfação de forma educada.
-	A IA vai gerar um plano de ação

Ordenar o plano de ação por classificação de risco

Saida Tabela:
ID da reclamação:
Data da reclamação:
Empresa:
Resumo da Reclamação:
Risco:
Plano de ação:


## Pré-processamento da base

In [ ]:
import sqlite3
import pandas as pd

In [2]:
conn = sqlite3.connect('./data/consumer_complaints.db')
cursor = conn.cursor()

# Imprimindo nome das colunas
cursor.execute(f"PRAGMA table_info(consumer_complaints);")
colunas = cursor.fetchall()
nomes_colunas = tuple(coluna[1] for coluna in colunas)
print(nomes_colunas)

# Visualizando algumas linhas da coluna
cursor.execute("SELECT * FROM consumer_complaints;")
for row in cursor.fetchall()[:10]:
    print(row)



('date_received', 'product', 'sub_product', 'issue', 'sub_issue', 'consumer_complaint_narrative', 'company_public_response', 'company', 'state', 'zipcode', 'tags', 'consumer_consent_provided', 'submitted_via', 'date_sent_to_company', 'company_response_to_consumer', 'timely_response', 'consumer_disputed?', 'complaint_id')
('08/30/2013', 'Mortgage', 'Other mortgage', 'Loan modification,collection,foreclosure', None, None, None, 'U.S. Bancorp', 'CA', '95993', None, None, 'Referral', '09/03/2013', 'Closed with explanation', 'Yes', 'Yes', 511074)
('08/30/2013', 'Mortgage', 'Other mortgage', 'Loan servicing, payments, escrow account', None, None, None, 'Wells Fargo & Company', 'CA', '91104', None, None, 'Referral', '09/03/2013', 'Closed with explanation', 'Yes', 'Yes', 511080)
('08/30/2013', 'Credit reporting', None, 'Incorrect information on credit report', 'Account status', None, None, 'Wells Fargo & Company', 'NY', '11764', None, None, 'Postal mail', '09/18/2013', 'Closed with explanation

Quantas reclamações que tem descrição da reclamação do cliente tem por empresa?  
##### 1 - Só consigo trabalhar com as reclamações que tem consentimento do cliente pra descrever a narrativa

In [3]:
pd_reclamacoes_totais = pd.read_sql_query("""select * from consumer_complaints where consumer_consent_provided like 'Consent provided'""", conn)
display(pd_reclamacoes_totais[:10])

date_received          product                            sub_product  \
0    03/19/2015  Debt collection  Other (i.e. phone, health club, etc.)   
1    03/19/2015    Consumer Loan                           Vehicle loan   
2    03/19/2015         Mortgage            Conventional fixed mortgage   
3    03/19/2015         Mortgage            Conventional fixed mortgage   
4    03/19/2015         Mortgage            Conventional fixed mortgage   
5    03/19/2015         Mortgage            Conventional fixed mortgage   
6    03/19/2015         Mortgage            Conventional fixed mortgage   
7    03/19/2015         Mortgage            Conventional fixed mortgage   
8    03/19/2015      Credit card                                   None   
9    03/19/2015    Consumer Loan                           Vehicle loan   

                                      issue      sub_issue  \
0     Cont'd attempts collect debt not owed  Debt was paid   
1                Managing the loan or lease           None   
2  Loan modification,collection,foreclosure           None   
3  Loan servicing, payments, escrow account           None   
4            Credit decision / Underwriting           None   
5  Loan servicing, payments, escrow account           None   
6  Loan modification,collection,foreclosure           None   
7  Loan servicing, payments, escrow account           None   
8                                     Other           None   
9              Taking out the loan or lease           None   

                        consumer_complaint_narrative  \
0  XXXX has claimed I owe them {$27.00} for XXXX ...   
1  Due to inconsistencies in the amount owed that...   
2  In XX/XX/XXXX my wages that I earned at my job...   
3  I have an open and current mortgage with Chase...   
4  XXXX was submitted XX/XX/XXXX. At the time I s...   
5  Experian is reporting my OPEN and CURRENT Mort...   
6  This complaint is against Wells Fargo Bank for...   
7  I spoke to XXXX of green tree representatives ...   
8  i opened XXXX Bank of America credit cards 15-...   
9  I applied for a loan with XXXX XXXX and had pu...   

                            company_public_response  \
0                                              None   
1                                              None   
2                                              None   
3                                              None   
4                                              None   
5  Company chooses not to provide a public response   
6                                              None   
7                                              None   
8  Company chooses not to provide a public response   
9                                              None   

                                 company state zipcode            tags  \
0          Diversified Consultants, Inc.    NY   121XX  Older American   
1                   M&T Bank Corporation    VA   221XX   Servicemember   
2                  Wells Fargo & Company    CA   946XX            None   
3                   JPMorgan Chase & Co.    CA   900XX  Older American   
4  Rushmore Loan Management Services LLC    CA   956XX  Older American   
5                               Experian    CA   900XX  Older American   
6                  Wells Fargo & Company    CA   931XX            None   
7                   Ditech Financial LLC    CA   948XX            None   
8                        Bank of America    OR   971XX            None   
9                Hyundai Capital America    MS   392XX            None   

  consumer_consent_provided submitted_via date_sent_to_company  \
0          Consent provided           Web           03/19/2015   
1          Consent provided           Web           03/19/2015   
2          Consent provided           Web           03/19/2015   
3          Consent provided           Web           03/19/2015   
4          Consent provided           Web           03/19/2015   
5          Consent provided           Web           03/25/2

Vamos ver quantas reclamações temos para trabalhar

In [4]:
pd_reclamacoes_totais.shape

(66807, 18)

Temos 18 colunas e 66807 reclamações. Destas reclamações, vamos agrupar a quantidade por empresas

In [5]:
cursor.execute("""select company, count(*)
                from consumer_complaints
                where consumer_complaint_narrative is not NULL
                group by company
                order by 2 desc;""")

pd_reclamacoes = pd_reclamacoes_totais[['company', 'complaint_id']].groupby('company').count().sort_values(by='complaint_id', ascending=False)
display(pd_reclamacoes[:10])

print(f'Total de reclamações: {pd_reclamacoes['complaint_id'].sum()}')

complaint_id
company                                             
Equifax                                         4195
Experian                                        3934
TransUnion Intermediate Holdings, Inc.          3863
Bank of America                                 3498
Wells Fargo & Company                           3075
Citibank                                        2789
JPMorgan Chase & Co.                            2586
Ocwen                                           1620
Capital One                                     1504
Synchrony Financial                             1375

Total de reclamações: 66807


Renomearei a coluna 'complaint_id' para 'qtd_reclamacoes'

In [6]:
pd_reclamacoes.rename(columns={'complaint_id': 'qtd_reclamacoes'}, inplace=True)
display(pd_reclamacoes[:10])

qtd_reclamacoes
company                                                
Equifax                                            4195
Experian                                           3934
TransUnion Intermediate Holdings, Inc.             3863
Bank of America                                    3498
Wells Fargo & Company                              3075
Citibank                                           2789
JPMorgan Chase & Co.                               2586
Ocwen                                              1620
Capital One                                        1504
Synchrony Financial                                1375

Quantas reclamações que o usuário descreveu a reclamação tiveram resposta por companhia?

In [7]:
pd_reclamacoes_respondidas = pd.read_sql_query("""select company, count(company_public_response) as reclamacoes_respondidas
                                                from consumer_complaints
                                                where consumer_complaint_narrative is not NULL
                                                and company_public_response is not NULL
                                                group by company
                                                order by 2 desc""", conn)
display(pd_reclamacoes_respondidas.head(10))

company  reclamacoes_respondidas
0                                Experian                     3934
1  TransUnion Intermediate Holdings, Inc.                     3605
2                         Bank of America                     3493
3                   Wells Fargo & Company                     2740
4                                Citibank                     2524
5                            U.S. Bancorp                     1077
6                    Ditech Financial LLC                      785
7                                     ERC                      513
8         Select Portfolio Servicing, Inc                      476
9                    SunTrust Banks, Inc.                      425

Preciso agora computar a taxa de reclamações respondidas = reclamações respondidas / total de reclamações com reclamação dos clientes  
Irei juntar as duas tabelas para ter essas informações para fazer o cálculo
Tem que ser **LEFT JOIN** porque pode ter empresa SEM RECLAMAÇÕES RESPONDIDAS

In [8]:
pd_reclamacoes_new = pd.merge(pd_reclamacoes, pd_reclamacoes_respondidas, on='company', how='left')
display(pd_reclamacoes_new.head(10))

company  qtd_reclamacoes  \
0                                 Equifax             4195   
1                                Experian             3934   
2  TransUnion Intermediate Holdings, Inc.             3863   
3                         Bank of America             3498   
4                   Wells Fargo & Company             3075   
5                                Citibank             2789   
6                    JPMorgan Chase & Co.             2586   
7                                   Ocwen             1620   
8                             Capital One             1504   
9                     Synchrony Financial             1375   

   reclamacoes_respondidas  
0                      NaN  
1                   3934.0  
2                   3605.0  
3                   3493.0  
4                   2740.0  
5                   2524.0  
6                      NaN  
7                     20.0  
8                      NaN  
9                    123.0

Repara que ao fazer o merge, algumas reclamações de certas companhias vieram com valor NaN.  
Ou seja, pode ser que não tenha reclamações respondidas destas empresas.  
Precisamos tratar esses nulos para fazer as contas

In [9]:
# Vou ver se valida a hipótese acima. Vou tentar pesquisar na outra tabela a empresa Equifax
pd_reclamacoes_respondidas.query("company == 'Equifax'")

Empty DataFrame
Columns: [company, reclamacoes_respondidas]
Index: []

Realmente a empresa Equifax não está na tabela de reclamações respondidas.

In [10]:
pd_reclamacoes_new.fillna(0, inplace=True)
display(pd_reclamacoes_new.head(10))

company  qtd_reclamacoes  \
0                                 Equifax             4195   
1                                Experian             3934   
2  TransUnion Intermediate Holdings, Inc.             3863   
3                         Bank of America             3498   
4                   Wells Fargo & Company             3075   
5                                Citibank             2789   
6                    JPMorgan Chase & Co.             2586   
7                                   Ocwen             1620   
8                             Capital One             1504   
9                     Synchrony Financial             1375   

   reclamacoes_respondidas  
0                      0.0  
1                   3934.0  
2                   3605.0  
3                   3493.0  
4                   2740.0  
5                   2524.0  
6                      0.0  
7                     20.0  
8                      0.0  
9                    123.0

Vou dar uma olhada no tipo das colunas. Essa nova coluna (reclamações respondidas) aparentemente não está no tipo correto (inteiro)  
Se não tiver, **irei fazer a conversão de tipo de função** 

In [11]:
pd_reclamacoes_new.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2296 entries, 0 to 2295
Data columns (total 3 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   company                  2296 non-null   object 
 1   qtd_reclamacoes          2296 non-null   int64  
 2   reclamacoes_respondidas  2296 non-null   float64
dtypes: float64(1), int64(1), object(1)
memory usage: 53.9+ KB


In [12]:
pd_reclamacoes_new['reclamacoes_respondidas'] = pd_reclamacoes_new['reclamacoes_respondidas'].astype(int)
display(pd_reclamacoes_new.head(10))

company  qtd_reclamacoes  \
0                                 Equifax             4195   
1                                Experian             3934   
2  TransUnion Intermediate Holdings, Inc.             3863   
3                         Bank of America             3498   
4                   Wells Fargo & Company             3075   
5                                Citibank             2789   
6                    JPMorgan Chase & Co.             2586   
7                                   Ocwen             1620   
8                             Capital One             1504   
9                     Synchrony Financial             1375   

   reclamacoes_respondidas  
0                        0  
1                     3934  
2                     3605  
3                     3493  
4                     2740  
5                     2524  
6                        0  
7                       20  
8                        0  
9                      123

Vamos computar a porcentagem e ordenar por ordem de maior taxa de resposta para menor taxa de resposta.

In [13]:
pd_reclamacoes_new['percentual_respondidas'] = round(pd_reclamacoes_new['reclamacoes_respondidas'] / pd_reclamacoes_new['qtd_reclamacoes'], 5)
display(pd_reclamacoes_new.sort_values(by=['percentual_respondidas', 'qtd_reclamacoes'], ascending=False).head(20))

company  qtd_reclamacoes  \
1                                  Experian             3934   
21          Select Portfolio Servicing, Inc              476   
31                      Ally Financial Inc.              326   
32                             USAA Savings              319   
42            Regions Financial Corporation              185   
46                                 Navy FCU              167   
48                               Sallie Mae              161   
60                                Loan Care              139   
70                    Focus Holding Company              115   
72   First National Collection Bureau, Inc.              114   
84            I.Q. Data International, Inc.               96   
86                    Stellar Recovery Inc.               94   
90                Enova International, Inc.               88   
92                          ProCollect, Inc               87   
106           Credit Acceptance Corporation               75   
108                     The CMI Group, Inc.               75   
109                 Penn Credit Corporation               73   
113         SquareTwo Financial Corporation               72   
118                              BMO Harris               69   
134        Amsher Collection Services, Inc.               57   

     reclamacoes_respondidas  percentual_respondidas  
1                       3934                     1.0  
21                       476                     1.0  
31                       326                     1.0  
32                       319                     1.0  
42                       185                     1.0  
46                       167                     1.0  
48                       161                     1.0  
60                       139                     1.0  
70                       115                     1.0  
72                       114                     1.0  
84                        96                     1.0  
86                        94                     1.0  
90                        88                     1.0  
92                        87                     1.0  
106                       75                     1.0  
108                       75                     1.0  
109                       73                     1.0  
113                       72                     1.0  
118                       69                     1.0  
134                       57                     1.0

Quando tenho uma base muito pequena de reclamações como visto acima, ter um percentual de 100% de taxa de resposta não se torna muito relevante para o problema.  
Então eu vou focar em **empresas que tem mais do que 400 reclamações** para ter uma amostra considerável de taxa de resposta.  

In [14]:
pd_reclamacoes_top400 = pd_reclamacoes_new.query('qtd_reclamacoes > 400')
display(pd_reclamacoes_top400)

company  qtd_reclamacoes  \
0                                  Equifax             4195   
1                                 Experian             3934   
2   TransUnion Intermediate Holdings, Inc.             3863   
3                          Bank of America             3498   
4                    Wells Fargo & Company             3075   
5                                 Citibank             2789   
6                     JPMorgan Chase & Co.             2586   
7                                    Ocwen             1620   
8                              Capital One             1504   
9                      Synchrony Financial             1375   
10                     Nationstar Mortgage             1372   
11                            U.S. Bancorp             1080   
12                    Ditech Financial LLC             1068   
13                 Navient Solutions, Inc.              997   
14                    Encore Capital Group              865   
15                                    Amex              740   
16     Portfolio Recovery Associates, Inc.              680   
17                   PayPal Holdings, Inc.              637   
18                                     ERC              622   
19                                Discover              560   
20               Empowerment Ventures, LLC              508   
21         Select Portfolio Servicing, Inc              476   
22                           PNC Bank N.A.              466   
23              TD Bank US Holding Company              457   
24                            Barclays PLC              427   
25                    SunTrust Banks, Inc.              426   

    reclamacoes_respondidas  percentual_respondidas  
0                         0                 0.00000  
1                      3934                 1.00000  
2                      3605                 0.93321  
3                      3493                 0.99857  
4                      2740                 0.89106  
5                      2524                 0.90498  
6                         0                 0.00000  
7                        20                 0.01235  
8                         0                 0.00000  
9                       123                 0.08945  
10                       43                 0.03134  
11                     1077                 0.99722  
12                      785                 0.73502  
13                       43                 0.04313  
14                        0                 0.00000  
15                        0                 0.00000  
16                        0                 0.00000  
17                      160                 0.25118  
18                      513                 0.82476  
19                        0                 0.00000  
20                        0                 0.00000  
21                      476                 1.00000  
22                        2                 0.00429  
23                        0                 0.00000  
24                      425                 0.99532  
25                      425                 0.99765

Agora sim temos grandes companhias que tem um sério problema de tempo de resposta para as reclamações. Vou ordenar novamente por percentual respondido

In [15]:
pd_reclamacoes_top400.sort_values(by=['percentual_respondidas', 'qtd_reclamacoes'], ascending=[True, False]).head(20)

company  qtd_reclamacoes  \
0                                  Equifax             4195   
6                     JPMorgan Chase & Co.             2586   
8                              Capital One             1504   
14                    Encore Capital Group              865   
15                                    Amex              740   
16     Portfolio Recovery Associates, Inc.              680   
19                                Discover              560   
20               Empowerment Ventures, LLC              508   
23              TD Bank US Holding Company              457   
22                           PNC Bank N.A.              466   
7                                    Ocwen             1620   
10                     Nationstar Mortgage             1372   
13                 Navient Solutions, Inc.              997   
9                      Synchrony Financial             1375   
17                   PayPal Holdings, Inc.              637   
12                    Ditech Financial LLC             1068   
18                                     ERC              622   
4                    Wells Fargo & Company             3075   
5                                 Citibank             2789   
2   TransUnion Intermediate Holdings, Inc.             3863   

    reclamacoes_respondidas  percentual_respondidas  
0                         0                 0.00000  
6                         0                 0.00000  
8                         0                 0.00000  
14                        0                 0.00000  
15                        0                 0.00000  
16                        0                 0.00000  
19                        0                 0.00000  
20                        0                 0.00000  
23                        0                 0.00000  
22                        2                 0.00429  
7                        20                 0.01235  
10                       43                 0.03134  
13                       43                 0.04313  
9                       123                 0.08945  
17                      160                 0.25118  
12                      785                 0.73502  
18                      513                 0.82476  
4                      2740                 0.89106  
5                      2524                 0.90498  
2                      3605                 0.93321

Fiquei curioso com as duas reclamações respondidas da PNC Bank N.A. (nome bem sugestivo... "Pau no..").  
Vou carregar a tabela inteira, filtrar por companhia e ver se encontro essas reclamações que os clientes descreveram a narrativa e a empresa respondeu 

In [16]:
display(pd.read_sql_query("""select complaint_id, company, date_received, product, sub_product, issue, sub_issue, consumer_complaint_narrative, company_public_response, submitted_via, timely_response
                            from consumer_complaints
                            where company like 'PNC%' 
                            and consumer_complaint_narrative is not NULL
                            and company_public_response is not NULL
                        """, conn))

complaint_id        company date_received      product  \
0       1347585  PNC Bank N.A.    04/26/2015     Mortgage   
1       1650258  PNC Bank N.A.    11/10/2015  Credit card   

                              sub_product  \
0  Conventional adjustable mortgage (ARM)   
1                                    None   

                                      issue sub_issue  \
0  Loan modification,collection,foreclosure      None   
1                                 Other fee      None   

                        consumer_complaint_narrative  \
0  I have submitted a request for release of a XX...   
1  please make sure you escalate to someone other...   

                            company_public_response submitted_via  \
0  Company chooses not to provide a public response           Web   
1  Company chooses not to provide a public response           Web   

  timely_response  
0             Yes  
1             Yes

Olha... as duas reclamações que a PNC resolveu responder publicamente foi justamente para dizer que "a empresa decidiu não responder publicamente".  
Em outras palavras, "Pau no..." no cliente.   
Eu já cancelaria a conta na hora.  
**Mas vamos fingir que eles ligam pra atendimento ao cliente e contrataram NÓS AQUI DA CONSULTORIA para melhorar a imagem deles através da IA.**   
Vou primeiramente coletar numa tabela todas as reclamações da firma sem resposta.

In [17]:
pd_reclamacoes_pnc = pd.read_sql_query("""select complaint_id, company, date_received, product, sub_product, issue, sub_issue, consumer_complaint_narrative, company_public_response, submitted_via, timely_response
                            from consumer_complaints
                            where company like 'PNC%' 
                            and consumer_complaint_narrative is not NULL
                            and company_public_response is NULL
                        """, conn)
display(pd_reclamacoes_pnc)

complaint_id        company date_received                  product  \
0         1290617  PNC Bank N.A.    03/19/2015            Consumer Loan   
1         1291299  PNC Bank N.A.    03/19/2015  Bank account or service   
2         1291414  PNC Bank N.A.    03/19/2015  Bank account or service   
3         1301711  PNC Bank N.A.    03/26/2015                 Mortgage   
4         1303393  PNC Bank N.A.    03/26/2015  Bank account or service   
..            ...            ...           ...                      ...   
459       1701859  PNC Bank N.A.    12/17/2015            Consumer Loan   
460       1775457  PNC Bank N.A.    02/05/2016          Money transfers   
461       1381042  PNC Bank N.A.    05/18/2015  Bank account or service   
462       1765721  PNC Bank N.A.    01/31/2016  Bank account or service   
463       1713158  PNC Bank N.A.    12/22/2015  Bank account or service   

                            sub_product  \
0                          Vehicle loan   
1    Cashing a check without an account   
2                      Checking account   
3           Conventional fixed mortgage   
4                      Checking account   
..                                  ...   
459                    Installment loan   
460        Domestic (US) money transfer   
461                    Checking account   
462          Other bank product/service   
463                    Checking account   

                                        issue sub_issue  \
0                  Managing the loan or lease      None   
1                    Deposits and withdrawals      None   
2     Account opening, closing, or management      None   
3    Loan servicing, payments, escrow account      None   
4     Account opening, closing, or management      None   
..                                        ...       ...   
459       Problems when you are unable to pay      None   
460                  Other transaction issues      None   
461     Problems caused by my funds being low      None   
462   Account opening, closing, or management      None   
463     Problems caused by my funds being low      None   

                          consumer_complaint_narrative  \
0    I have a car loan with PNC bank. In XXXX of 20...   
1    I have a contract with a guy whom has an accou...   
2    I signed up for a PNC account on XXXX XXXX 201...   
3    PNC Mortgage forces me to pay {$7.00} fee each...   
4    I opened a new checking account [ XXXX ] with ...   
..                                                 ...   
459  I intended to buy a home with a mortgage in 20...   
460  My daughter set me up as a payee in PNC 's " p...   
461  My bank statement reflects that on the XX/XX/X...   
462  On XXXX/XXXX/2015, PNC Bank accessed my XXXX p...   
463  My bank continues to charge me overdraft fees ...   

    company_public_response submitted_via timely_response  
0                      None           Web             Yes  
1                      None           Web             Yes  
2                      None           Web             Yes  
3                      None           Web             Yes  
4                      None           Web             Yes  
..                      ...           ...             ...  
459                    None           Web             Yes  
460                    None           Web             Yes  
461                    None           Web             Yes  
462                    None           Web             Yes  
463                    None           Web             Yes  

[464 rows x 11 columns]

Vou remover as chaves { e } do texto para não dar problema no Prompt Template. Para isso vou usar uma função apply para aplicar em toda coluna da reclamação do cliente

In [80]:
def remove_chaves(text):
    return text.replace('{', '').replace('}', '')

pd_reclamacoes_pnc['consumer_complaint_narrative'] = pd_reclamacoes_pnc['consumer_complaint_narrative'].apply(remove_chaves)
display(pd_reclamacoes_pnc)

complaint_id        company date_received                  product  \
0         1290617  PNC Bank N.A.    03/19/2015            Consumer Loan   
1         1291299  PNC Bank N.A.    03/19/2015  Bank account or service   
2         1291414  PNC Bank N.A.    03/19/2015  Bank account or service   
3         1301711  PNC Bank N.A.    03/26/2015                 Mortgage   
4         1303393  PNC Bank N.A.    03/26/2015  Bank account or service   
..            ...            ...           ...                      ...   
459       1701859  PNC Bank N.A.    12/17/2015            Consumer Loan   
460       1775457  PNC Bank N.A.    02/05/2016          Money transfers   
461       1381042  PNC Bank N.A.    05/18/2015  Bank account or service   
462       1765721  PNC Bank N.A.    01/31/2016  Bank account or service   
463       1713158  PNC Bank N.A.    12/22/2015  Bank account or service   

                            sub_product  \
0                          Vehicle loan   
1    Cashing a check without an account   
2                      Checking account   
3           Conventional fixed mortgage   
4                      Checking account   
..                                  ...   
459                    Installment loan   
460        Domestic (US) money transfer   
461                    Checking account   
462          Other bank product/service   
463                    Checking account   

                                        issue sub_issue  \
0                  Managing the loan or lease      None   
1                    Deposits and withdrawals      None   
2     Account opening, closing, or management      None   
3    Loan servicing, payments, escrow account      None   
4     Account opening, closing, or management      None   
..                                        ...       ...   
459       Problems when you are unable to pay      None   
460                  Other transaction issues      None   
461     Problems caused by my funds being low      None   
462   Account opening, closing, or management      None   
463     Problems caused by my funds being low      None   

                          consumer_complaint_narrative  \
0    I have a car loan with PNC bank. In XXXX of 20...   
1    I have a contract with a guy whom has an accou...   
2    I signed up for a PNC account on XXXX XXXX 201...   
3    PNC Mortgage forces me to pay $7.00 fee each t...   
4    I opened a new checking account [ XXXX ] with ...   
..                                                 ...   
459  I intended to buy a home with a mortgage in 20...   
460  My daughter set me up as a payee in PNC 's " p...   
461  My bank statement reflects that on the XX/XX/X...   
462  On XXXX/XXXX/2015, PNC Bank accessed my XXXX p...   
463  My bank continues to charge me overdraft fees ...   

    company_public_response submitted_via timely_response  
0                      None           Web             Yes  
1                      None           Web             Yes  
2                      None           Web             Yes  
3                      None           Web             Yes  
4                      None           Web             Yes  
..                      ...           ...             ...  
459                    None           Web             Yes  
460                    None           Web             Yes  
461                    None           Web             Yes  
462                    None           Web             Yes  
463                    None           Web             Yes  

[464 rows x 11 columns]

### Hora de botar a IA pra trabalhar

In [ ]:
import os
from langchain.schema import StrOutputParser
from langchain_openai import ChatOpenAI
from langchain_core.prompts import PromptTemplate
from dotenv import load_dotenv

Pegarei uma amostra de 5 reclamações únicas aleatórias

In [ ]:
# Função sample coleta uma amostra aleatória de um DataFrame.
# Defini um random_state para toda vez que rodar, selecionar a mesma amostra. 
# O Replace = False garante que não haverá repetição de linhas na amostra.

df_sample = pd_reclamacoes_pnc.sample(5, random_state=42, replace=False)
display(df_sample)

complaint_id        company date_received                  product  \
9         1294663  PNC Bank N.A.    03/21/2015                 Mortgage   
30        1312581  PNC Bank N.A.    04/02/2015  Bank account or service   
39        1346696  PNC Bank N.A.    04/24/2015  Bank account or service   
443       1800494  PNC Bank N.A.    02/23/2016            Consumer Loan   
266       1576383  PNC Bank N.A.    09/23/2015                 Mortgage   

                     sub_product                                     issue  \
9    Conventional fixed mortgage  Loan modification,collection,foreclosure   
30              Checking account     Problems caused by my funds being low   
39              Checking account   Account opening, closing, or management   
443                 Vehicle loan                Managing the loan or lease   
266                 FHA mortgage              Settlement process and costs   

    sub_issue                       consumer_complaint_narrative  \
9        None  PNC is servicing my mortgage. I filed for bank...   
30       None  I bank with PNC Bank. I have for 10 years. Rec...   
39       None   PNC bank is falsely submitting overdraft fees \n   
443      None  I have a vehicle financed of which i paid a to...   
266      None  I am the XXXX for buyers of XXXX XXXX XXXX, XX...   

    company_public_response submitted_via timely_response  
9                      None           Web             Yes  
30                     None           Web             Yes  
39                     None           Web             Yes  
443                    None           Web             Yes  
266                    None           Web             Yes

Configuraremos o modelo de IA para executar as tarefas do desafio.   
Precisamos carregar as chaves de acesso utilizando load_dotenv().  
**Não esqueça de criar um arquivo texto .env com as credenciais necessárias!!!**

In [ ]:
load_dotenv()

openai_key = os.getenv('OPENAI_API_KEY')

llm = ChatOpenAI(temperature=0,
                 model_name="gpt-4o-mini")


Criando classe com as 5 funções que utilizará a IA para processamento das reclamações

In [96]:
class ComplainSolverIA():
    def __init__(self, llm_model):
        self.llm = llm_model

    def function_ia_translator(self, user_message):
        """Esta função tem o objetivo de traduzir o prompt do inglês para português""" 

        prompt = f"""Você é um especialista em inglês. Seu objetivo é traduzir o texto abaixo para o português. 
                            \n\n{user_message}"""
        
        prompt_template = PromptTemplate(input_variables=["user_message"],
                                        template=prompt)
        
        chain = prompt_template | self.llm | StrOutputParser()

        return chain.invoke(input={"user_message": user_message})


    def function_ia_summary(self, user_message):
        """Esta função tem o objetivo de sumarizar a reclamação do cliente""" 

        prompt = f"""Você é encerregado em resumir em até 100 caracteres o resumo da reclamação do usuário para 
                    a empresa para passar a reclamação para o departamento de avaliação de risco.  
                            \n\n{user_message}"""
        
        prompt_template = PromptTemplate(input_variables=["user_message"],
                                        template=prompt)
        
        chain = prompt_template | self.llm | StrOutputParser()

        return chain.invoke(input={"user_message": user_message})
    
    def function_analysis_sentiment(self, user_message):
        """Esta função tem o objetivo de avaliar o grau de sentimento da mensagem.""" 

        prompt = f"""Você fará o papel do usuário que está reclamando. Sua missão é analisar o conteúdo da reclamação
                    e classificar o sentimento predominante da mensagem. 
                    Reclamação do cliente:\n
                    {user_message}"""
        
        prompt_template = PromptTemplate(input_variables=["user_message"],
                                        template=prompt)
        
        chain = prompt_template | self.llm | StrOutputParser()

        return chain.invoke(input={"user_message": user_message})
    
    def function_risk_evaluator(self, user_message):
        """Esta função tem o objetivo de avaliar o grau de risco do cliente com relação a reclamação.""" 

        prompt = f"""Você é um funcionário de avaliação de risco de uma empresa do segmento financeiro.
                    Sua função é a partir da reclamação do usuário, classificar o grau de risco de deixar a empresa em ALTO, MEDIO e BAIXO.
                    Segue o contexto para se levar em consideração na avaliação do risco do cliente:
                    \n
                    Risco Alto: Palavras que ameaça diretamente a empresa com processo, advogado, linguagem 
                    agressiva ou tóxica. Ameaça deixar de utilizar os produtos da empresa.
                    Risco Médio: Cliente não ameaça diretamente a empresa mas utiliza algumas palavras agressivas 
                    demonstrando sinal de estresse mas não dá indícios de deixar de consumir algum produto da empresa.
                    Risco Baixo: Cliente faz a reclamação sem ameaças a empresa e demonstra insatisfação de forma educada.
                    \n
                    Repare o tom de sentimento da mensagem do cliente. Se ele apresentar sinais claros de desespero, frustração
                    ou estresse, o risco desse cliente é ALTO porque ele não agirá mais com o racional. 
                    \n
                    Dito isto, retorne em apenas e unicamente com a string "ALTO", "MEDIO" ou "BAIXO" a reclamação do cliente.
                    \n
                    Reclamação do cliente:\n
                    {user_message}"""
        
        prompt_template = PromptTemplate(input_variables=["user_message"],
                                        template=prompt)
        
        chain = prompt_template | self.llm | StrOutputParser()

        return chain.invoke(input={"user_message": user_message})
    
    def function_came_to_solve_everything(self, user_message):
        """Esta função tem o objetivo de avaliar o grau de risco do cliente com relação a reclamação.""" 

        prompt = f"""Você é um especialistas em relações públicas. Sua missão é gerar uma mensagem personalizada em até
                    200 caracteres para enviar ao cliente. A mensagem deve contemplar:
                    1 - O problema especificado
                    2 - A preocupação da empresa em resolvê-lo
                    3 - Pedido de desculpas APENAS e UNICAMENTE se a EMPRESA foi a CULPADA pelo problema.
                    Reclamação do cliente:\n
                    {user_message}
                    Saída esperada:
                    Mensagem para o Cliente:"""
        
        prompt_template = PromptTemplate(input_variables=["user_message"],
                                        template=prompt)
        
        chain = prompt_template | self.llm | StrOutputParser()

        return chain.invoke(input={"user_message": user_message})



Criando a função que gera essa tabela de reclamação

In [105]:
def complains_table(df_table, llm_model):
    """
        Esta função monta a tabela final com as reclamações dos clientes, resumo da reclamação, sentimento da reclamação,
        risco do cliente e mensagem de ação para o cliente.
    """
    complaint_id = df_table['complaint_id'].values
    complaint_date = df_table['date_received'].values
    company = df_table['company'].values
    reclamacao = []
    resumo_reclamacao = []
    sentimento_reclamacao = []
    risco_cliente = []
    mensagem_acao = []

    ia = ComplainSolverIA(llm_model)

    for _, row in df_table.iterrows():
        print(f"Processando reclamação {row['complaint_id']}")
        reclamacao.append(ia.function_ia_translator(row['consumer_complaint_narrative']))
        resumo_reclamacao.append(ia.function_ia_summary(row['consumer_complaint_narrative']))
        sentimento_reclamacao.append(ia.function_analysis_sentiment(row['consumer_complaint_narrative']))
        risco_cliente.append(ia.function_risk_evaluator(row['consumer_complaint_narrative']))
        mensagem_acao.append(ia.function_came_to_solve_everything(row['consumer_complaint_narrative']))

    return pd.DataFrame({'id_reclamacao': complaint_id,
                         'data_reclamacao': complaint_date,
                         'empresa': company,
                         'reclamacao_cliente': reclamacao,
                         'resumo_reclamacao_cliente': resumo_reclamacao,
                         'sentimento_reclamacao': sentimento_reclamacao,
                         'risco_cliente': risco_cliente,
                         'mensagem_acao': mensagem_acao})

Exibindo a tabela final

In [106]:
df_final = complains_table(df_sample, llm)
display(df_final)

Processando reclamação 1294663
Processando reclamação 1312581
Processando reclamação 1346696
Processando reclamação 1800494
Processando reclamação 1576383


id_reclamacao data_reclamacao        empresa  \
0        1294663      03/21/2015  PNC Bank N.A.   
1        1312581      04/02/2015  PNC Bank N.A.   
2        1346696      04/24/2015  PNC Bank N.A.   
3        1800494      02/23/2016  PNC Bank N.A.   
4        1576383      09/23/2015  PNC Bank N.A.   

                                  reclamacao_cliente  \
0  A PNC está administrando a minha hipoteca. Eu ...   
1  Eu sou cliente do PNC Bank há 10 anos. Recente...   
2  O banco PNC está cobrando taxas de descoberto ...   
3  Claro, aqui está a tradução do texto para o po...   
4  Eu sou o XXXX para os compradores de XXXX XXXX...   

                           resumo_reclamacao_cliente  \
0  PNC atrasou venda pós-falência, gerando multas...   
1  PNC desconta valores indevidos e cobra taxas e...   
2  PNC Bank está cobrando taxas de overdraft inde...   
3  Financiamento pago integralmente, cobrança ind...   
4  Atrasos na avaliação impedem fechamento, apesa...   

                               sentimento_reclamacao risco_cliente  \
0  Após analisar o conteúdo da reclamação, o sent...          ALTO   
1  **Análise de Sentimento:**\n\nO sentimento pre...          ALTO   
2  **Sentimento Predominante:** Negativo\n\n**Aná...         MEDIO   
3  A reclamação do cliente expressa predominantem...          ALTO   
4  **Sentimento predominante:** Negativo\n\n**Aná...         MEDIO   

                                       mensagem_acao  
0  Lamentamos profundamente os transtornos enfren...  
1  Mensagem para o Cliente:\n\nLamentamos os inco...  
2  Prezado(a),\n\nPedimos desculpas pelos encargo...  
3  Olá [Cliente], lamentamos pelo transtorno com ...  
4  **Mensagem para o Cliente:**\n\nLamentamos pel...

Exportando para csv

In [108]:
df_final.to_csv('reclamacoes_final.csv', index=False, encoding='utf-8')